File: windows/Max.Core/Loops/LoopTool.cs
using System.Text;
using System.Text.Json.Nodes;

namespace Max.Core;

/// <summary>Lets Max create, list, and delete its own scheduled loops.</summary>
public sealed class LoopTool : IMaxTool
{
    public ToolSpec Spec => new(
        "loop",
        "Manage scheduled autonomous runs ('loops'). action=create schedules a recurring task; " +
        "action=list shows existing loops; action=delete removes one by id. For create, give a clear " +
        "prompt describing the task. schedule is 'every' (with interval_minutes), 'daily' (with time HH:mm), " +
        "or 'once' (with at, an ISO-8601 UTC timestamp).",
        new JsonObject
        {
            ["type"] = "object",
            ["properties"] = new JsonObject
            {
                ["action"] = new JsonObject { ["type"] = "string", ["enum"] = new JsonArray("create", "list", "delete") },
                ["name"] = new JsonObject { ["type"] = "string" },
                ["prompt"] = new JsonObject { ["type"] = "string" },
                ["schedule"] = new JsonObject { ["type"] = "string", ["enum"] = new JsonArray("every", "daily", "once") },
                ["interval_minutes"] = new JsonObject { ["type"] = "integer" },
                ["time"] = new JsonObject { ["type"] = "string", ["description"] = "HH:mm local, for daily" },
                ["at"] = new JsonObject { ["type"] = "string", ["description"] = "ISO-8601 UTC, for once" },
                ["id"] = new JsonObject { ["type"] = "string", ["description"] = "loop id, for delete" },
            },
            ["required"] = new JsonArray("action"),
        });

    public string Summary(JsonObject input) => $"loop {(string?)input["action"] ?? "?"}";

    public Task<ToolOutcome> ExecuteAsync(JsonObject input, CancellationToken ct)
    {
        var action = (string?)input["action"] ?? "list";
        switch (action)
        {
            case "create":
            {
                var loop = new LoopDef
                {
                    Name = (string?)input["name"] ?? "Untitled loop",
                    Prompt = (string?)input["prompt"] ?? "",
                    Schedule = Enum.TryParse<LoopSchedule>((string?)input["schedule"], true, out var s) ? s : LoopSchedule.Every,
                    IntervalMinutes = (int?)input["interval_minutes"] ?? 60,
                    Time = (string?)input["time"] ?? "09:00",
                    AtUtc = DateTime.TryParse((string?)input["at"], null, System.Globalization.DateTimeStyles.AdjustToUniversal, out var at) ? at : null,
                };
                if (string.IsNullOrWhiteSpace(loop.Prompt))
                    return Task.FromResult(ToolOutcome.Fail("a loop needs a prompt describing the task"));
                LoopStore.Upsert(loop);
                return Task.FromResult(ToolOutcome.Ok($"Created loop '{loop.Name}' (id {loop.Id}), schedule={loop.Schedule}."));
            }
            case "delete":
            {
                var id = (string?)input["id"];
                if (string.IsNullOrWhiteSpace(id)) return Task.FromResult(ToolOutcome.Fail("delete needs an id"));
                LoopStore.Delete(id!);
                return Task.FromResult(ToolOutcome.Ok($"Deleted loop {id}."));
            }
            default:
            {
                var all = LoopStore.Load();
                if (all.Count == 0) return Task.FromResult(ToolOutcome.Ok("No loops scheduled."));
                var sb = new StringBuilder();
                foreach (var l in all)
                    sb.AppendLine($"- [{l.Id}] {l.Name} — {l.Schedule}" +
                        (l.Schedule == LoopSchedule.Every ? $" {l.IntervalMinutes}m" :
                         l.Schedule == LoopSchedule.Daily ? $" @ {l.Time}" : $" @ {l.AtUtc:u}") +
                        (l.Enabled ? "" : " (done)"));
                return Task.FromResult(ToolOutcome.Ok(sb.ToString().TrimEnd()));
            }
        }
    }
}

File: Sources/Max/Core/Loops/LoopTool.swift
import Foundation

/// The agent-facing scheduling tool — OpenClaw's `cron` tool, simplified.
/// Lets the user say "check X every hour" and have Max wire it up itself.
struct LoopTool: MaxTool {
    var spec: ToolSpec {
        ToolSpec(
            name: "loop",
            description: """
            Manage recurring loops — autonomous scheduled runs of Max. Use when the user wants \
            something done repeatedly or later ("every morning", "every 30 minutes", "tonight at 9"). \
            Each run executes `prompt` as a fresh Max session with full tool access; results are \
            delivered as macOS notifications and stored in the Loops panel. \
            Actions: add (name, prompt, and one of every_minutes / daily_at / once_at), \
            list, remove (name), run_now (name).
            """,
            inputSchema: [
                "type": "object",
                "properties": [
                    "action": [
                        "type": "string",
                        "enum": ["add", "list", "remove", "run_now"],
                    ],
                    "name": ["type": "string", "description": "Short loop name, e.g. 'morning-brief'"],
                    "prompt": ["type": "string", "description": "What Max should do on each run"],
                    "every_minutes": ["type": "integer", "description": "Run every N minutes"],
                    "daily_at": ["type": "string", "description": "Run daily at HH:mm (24h), e.g. '08:30'"],
                    "once_at": ["type": "string", "description": "Run once at ISO-8601 time, e.g. '2026-06-12T21:00:00Z'"],
                ],
                "required": ["action"],
            ]
        )
    }

    func summary(input: [String: Any]) -> String {
        let action = input["action"] as? String ?? "?"
        let name = input["name"] as? String ?? ""
        return "loop \(action) \(name)".trimmingCharacters(in: .whitespaces)
    }

    func execute(input: [String: Any]) async -> ToolOutcome {
        let action = input["action"] as? String ?? ""
        switch action {
        case "add": return await add(input)
        case "list": return await list()
        case "remove": return await remove(input)
        case "run_now": return await runNow(input)
        default: return .fail("unknown action '\(action)'")
        }
    }

    private func parseSchedule(_ input: [String: Any]) -> LoopSchedule? {
        if let minutes = input["every_minutes"] as? Int, minutes >= 1 {
            return .everyMinutes(minutes)
        }
        if let daily = input["daily_at"] as? String {
            let parts = daily.split(separator: ":").compactMap { Int($0) }
            if parts.count == 2, (0..<24).contains(parts[0]), (0..<60).contains(parts[1]) {
                return .dailyAt(hour: parts[0], minute: parts[1])
            }
        }
        if let once = input["once_at"] as? String {
            let formatter = ISO8601DateFormatter()
            if let date = formatter.date(from: once) { return .once(date) }
            formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            if let date = formatter.date(from: once) { return .once(date) }
        }
        return nil
    }

    private func add(_ input: [String: Any]) async -> ToolOutcome {
        guard let name = input["name"] as? String, !name.isEmpty,
              let prompt = input["prompt"] as? String, !prompt.isEmpty else {
            return .fail("add requires `name` and `prompt`")
        }
        guard let schedule = parseSchedule(input) else {
            return .fail("add requires one of every_minutes, daily_at (HH:mm) or once_at (ISO-8601)")
        }
        let loop = MaxLoop(name: name, prompt: prompt, schedule: schedule)
        await MainActor.run {
            LoopStore.shared.remove(id: name) // replace same-named loop
            LoopStore.shared.add(loop)
        }
        return .ok("Loop '\(name)' scheduled — \(schedule.displayName).")
    }

    private func list() async -> ToolOutcome {
        let loops = await MainActor.run { LoopStore.shared.loops }
        if loops.isEmpty { return .ok("No loops scheduled.") }
        let lines = loops.map { loop in
            let status = loop.enabled ? "on" : "off"
            let last = loop.lastSummary.map { " | last: \($0.prefix(80))" } ?? ""
            return "- \(loop.name) [\(status)] \(loop.schedule.displayName): \(loop.prompt.prefix(80))\(last)"
        }
        return .ok(lines.joined(separator: "\n"))
    }

    private func remove(_ input: [String: Any]) async -> ToolOutcome {
        guard let name = input["name"] as? String, !name.isEmpty else {
            return .fail("remove requires `name`")
        }
        let existed = await MainActor.run { () -> Bool in
            let had = LoopStore.shared.loops.contains { $0.name == name || $0.id == name }
            LoopStore.shared.remove(id: name)
            return had
        }
        return existed ? .ok("Loop '\(name)' removed.") : .fail("No loop named '\(name)'.")
    }

    private func runNow(_ input: [String: Any]) async -> ToolOutcome {
        guard let name = input["name"] as? String,
              let loop = await MainActor.run(body: { LoopStore.shared.loops.first { $0.name == name || $0.id == name } })
        else { return .fail("No loop with that name.") }
        await MainActor.run { LoopScheduler.shared.runNow(loop) }
        return .ok("Loop '\(loop.name)' kicked off in the background.")
    }
}
